### PR TITLE
feat: Configurable ArgoCD binary download links on Help page. Fixes #7698

### LINKED ISSUE #7698: Configurable download links on Help page
# Summary

Argo CD image includes 3 binaries: linux, mac and window for x64 architecture.  The binaries can be downloaded by end-users using `download/argocd-linux-amd64` etc URLs. The URLs are discoverable using `/help` page. 

Instead of pre-bundling all kinds of binaries, it is proposed to keep only linux amd64/arm (depends on image type) and remove mac/windows. Support configuring downloadable binaries using Argo CD settings

# Motivation

Each `argocd` binary is ~100mb. With ARM we will end up having 6 binaries which is too much and not required to everyone. 

# Proposal

* The default image should have only `argocd-linux-amd64` or `argocd-linux-arm` binary as well as `/help` page should have only one link
* User should be able to configure additional binaries using `argocd-cm` ConfigMap:

```yaml
data
  # help.download. is a prefix; argocd-windows-amd64 is a downloaded file name and value is a URL
  help.download.argocd-windows-amd64: https://github.com/argoproj/argo-cd/releases/download/v2.1.6/argocd-windows-amd64.exe
  # value can be local path as well
  help.download.argocd-darwin-amd64: /additional_tools/argocd-darwin-amd64 
  # value can be prefixed with optional title
  help.download.argocd-darwin-arm: Windows ARM|/additional_tools/argocd-darwin-arm
```

### PR DESCRIPTION
**Pull request by terrytangyuan**
Fixes https://github.com/argoproj/argo-cd/issues/7698.

image
image

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: ui/src/app/help/components/help.tsx [hunk 1]
```diff
@@ -1,40 +1,78 @@
...
+                                    </div>
+                                    <div className='columns large-4 small-6'>
+                                        <div className='help-box'>
+                                            <p>Want to download the CLI tool?</p>
+                                            <a href={`download/argocd-linux-${process.env.HOST_ARCH}`} className='user-info-panel-buttons argo-button argo-button--base'>
+                                                <i className='fab fa-linux' /> Linux
+                                            </a>
+                                            &nbsp;
+                                            {binaryUrls.hasOwnProperty('darwin-amd64') && (
+                                                <a href={`${binaryUrls['darwin-amd64']}`} className='user-info-panel-buttons argo-button argo-button--base'>
```
Comment (terrytangyuan) [inline ui/src/app/help/components/help.tsx:44]: Check whether this works with real HTTP URLs.

Comment (terrytangyuan) [reply ui/src/app/help/components/help.tsx:44]: Yes this works (see bottom left of the screenshot). The download also pops up correctly when I click the link.

image

**Review by jessesuen — APPROVED** [review]
From backend perspective, it LGTM.

@alexmt any issue with the frontend?

File: util/settings/settings.go [hunk 2]
```diff
@@ -1061,6 +1068,16 @@ func (mgr *SettingsManager) ensureSynced(forceResync bool) error {
 	return mgr.initialize(ctx)
 }
 
+func getDownloadBinaryUrlsFromConfigMap(argoCDCM *apiv1.ConfigMap) map[string]string {
+	binaryUrls := map[string]string{}
+	for _, archType := range []string{"darwin-amd64", "darwin-arm64", "windows-amd64"} {
```
Comment (alexmt) [inline util/settings/settings.go:1073]: Users might want to add arm/amd linux . I think list should include `linux-arm` and `linux-amd64`

**Review by alexmt — CHANGES_REQUESTED** [review]
Thank you @terrytangyuan . Added couple suggestions for API and UI

File: util/settings/settings.go (hunk 2 above)
Comment (terrytangyuan) [reply util/settings/settings.go]: Added

File: ui/src/app/help/components/help.tsx [hunk 3]
```diff
@@ -1,40 +1,78 @@
...
+                                        </div>
+                                    </div>
+                                    <div className='columns large-4 small-6'>
+                                        <div className='help-box'>
+                                            <p>Want to download the CLI tool?</p>
+                                            <a href={`download/argocd-linux-${process.env.HOST_ARCH}`} className='user-info-panel-buttons argo-button argo-button--base'>
+                                                <i className='fab fa-linux' /> Linux
+                                            </a>
+                                            &nbsp;
+                                            {binaryUrls.hasOwnProperty('darwin-amd64') && (
```
[the remark this answers is not in the document]
Comment (terrytangyuan) [reply ui/src/app/help/components/help.tsx:43]: I tried refactoring this in https://github.com/argoproj/argo-cd/pull/7755/commits/5e93d0080071ec4ea823d744dd0e59f3c874d6b0 but got stuck with some issues. 

Though I was able to fix the page crushing issue very easily. Would you mind helping with refactoring the typescript code later as I have zero experience and don't have much bandwidth to debug further?

[the remark this answers is not in the document]
Comment (alexmt) [reply ui/src/app/help/components/help.tsx:49]: Sorry for long response. Thanks for fixing the crash. I think it is ok to keep it as is - refactoring is optional.

**Review by alexmt — APPROVED** [review]

**Comment by alexmt** [comment]
@terrytangyuan , I just realized the change should be in upgrading instructions. Can you please document the change in `docs/operator-manual/upgrading/2.2-2.3.md` ?

**Comment by terrytangyuan** [comment]
> @terrytangyuan , I just realized the change should be in upgrading instructions. Can you please document the change in `docs/operator-manual/upgrading/2.2-2.3.md` ?

Good call. I'll send a PR today.

**Pull request merged**